ion-schema/src/nfa.rs

```diff
@@ -4,135 +4,219 @@ use crate::types::TypeValidator;
 }
 
 impl NfaRun {
+    pub fn new(current_state_id: StateId, visits_to_current_state: usize) -> Self {
+        Self {
+            state_id: current_state_id,
+            state_visits: visits_to_current_state,
+        }
+    }
+}
+
+/// This is a context which will be used while validating an Ion value for `ordered_elements` constraint using its NFA
+#[derive(Debug, Clone)]
+pub struct NfaEvaluation {
+    pub(crate) visits: HashSet<NfaRun>,
+    pub(crate) nfa: Rc<Nfa>,
+}
+
+impl NfaEvaluation {
     pub fn new(nfa: Rc<Nfa>) -> Self {
         Self {
             visits: {
-                let mut visits = HashMap::new();
-                visits.insert(0, 1);
+                let mut visits = HashSet::new();
+                if !nfa.get_final_states().is_empty() {
+                    visits.insert(NfaRun::new(0, 0));
+                }
                 visits
             },
             nfa,
         }
     }
 
-    pub fn has_final_state(&self) -> bool {
-        self.visits
-            .iter()
-            .any(|(id, _)| self.nfa.states[*id] == State::Final)
+    pub fn has_final_state(&self, type_store: &TypeStore) -> bool {
+        if self.nfa.get_final_states().is_empty() && self.visits.is_empty() {
+            return true;
+        }
+
+        self.visits.iter().any(|nfa_run| {
+            self.nfa.get_final_states().iter().any(|fs| {
+                fs.state_id == nfa_run.state_id
+                    && nfa_run.state_visits >= fs.min
+                    && nfa_run.state_visits <= fs.max
+            })
+        })
     }
 
-    pub fn transition(
-        &mut self,
-        event: Option<Event>,
-        type_store: &TypeStore,
-    ) -> HashMap<usize, usize> {
-        let mut new_states = HashMap::new();
-        for (transition_from_state_id, visits) in self.visits.iter() {
-            let can_exit = self.nfa.states[*transition_from_state_id].can_exit(visits);
-            let transition_to_states: &HashSet<Transition> =
-                self.nfa.transitions.get(transition_from_state_id).unwrap();
-            self.to_states_transitions(
-                transition_from_state_id,
-                transition_to_states,
-                can_exit,
-                visits,
-                event.to_owned(),
-                type_store,
-                &mut new_states,
-            );
+    pub fn validate_ordered_elements(&mut self, elements: Vec<Element>, type_store: &TypeStore) {
+        let mut elements_iter = elements.iter().peekable();
+        // use nfa_evaluation for validation
+        while let Some(element) = elements_iter.next() {
+            let mut next_states = HashSet::new();
+            for nfa_run in self.visits.iter() {
+                let transitions: HashSet<Transition> = self.nfa.get_transitions(nfa_run.state_id);
+                self.evaluate_transitions(
+                    nfa_run,
+                    transitions,
+                    element,
+                    &mut elements_iter,
```

Comment:
Curious why we're passing through a `Peekable<Iter<Element>>` instead of just the `&[Element]`.

Reply:
In order to move all information to transitions, I had to change the way can_exit check was working. Previously with can_exit check I would have to verify it for each element for given Ion Elements. But now I am using a peekable iterator to just satisfy the minimum occurrence at once and then exit that state. This removes the need to check for can_exit everytime for an element.